### PR TITLE
I001 Adding profile picture

### PR DESCRIPTION
**Pull request by sdagn1**
When logged in, a simple Profile Page is available to the user, which is available on the navbar.
Screenshot 2024-09-12 at 10 27 21

Clicking on the Update Profile Picture will enable their webcam and ask for a photo to be taken.
Screenshot 2024-09-12 at 10 27 31

The photo will be converted to SVG and displayed as a thumbnail.
Screenshot 2024-09-12 at 10 27 41

Then clicking accept will send the compressed SVG Profile Picture to the API, and then retrieve it.
Screenshot 2024-09-12 at 10 27 50

Halloween Theme:
Screenshot 2024-09-12 at 15 20 26
Screenshot 2024-09-12 at 15 20 36
Screenshot 2024-09-12 at 15 20 43

Screenshot 2024-09-12 at 15 21 15
Screenshot 2024-09-12 at 15 20 56

### CODE REVIEW

**Review by Adnanhoq — APPROVED** [review]

File: src/scripts/imageScaling.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+            ctx.drawImage(this, 0, 0, wantedWidth, wantedHeight);
+
+            var dataURI = canvas.toDataURL();
+
+            resolve(dataURI);
+        };
+        img.src = datas;
+
+    })
+}//
```
Comment (Leebee1392) [inline src/scripts/imageScaling.js:21]: Is there meant to be a // ?

File: views/profilePicture.njk [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        <video id="webcam" autoplay playsinline class="rounded-circle" width="500" height="500"></video>
+        <canvas id="canvas" class="d-none"></canvas>
+        <div class="d-flex w-100 justify-content-center">
+            <button class="mt-3 btn btn-secondary w-50 mb-5" id="takePicture">Take Picture</button>
+        </div>
+</div>
+</div>
+
+<script>
+const webcamElement = document.getElementById('webcam');
```
Comment (Leebee1392) [inline views/profilePicture.njk:44]: Could this be in its own js file?

File: views/layouts/main.njk [hunk 3]
```diff
@@ -16,6 +16,10 @@
  
     {% endblock %}
 
+    {% block scripts %}
```
Comment (Leebee1392) [inline views/layouts/main.njk:19]: Is this needed?

File: views/components/navbar.njk [hunk 4]
```diff
@@ -26,8 +27,13 @@
     </div>
     <div class="sign-in-container d-flex justify-content-center align-items-center gap-2">
 
-
+        {% if profile %}
+        <div class="rounded-circle overflow-hidden">
+        {{profile | safe}}
```
Comment (Leebee1392) [inline views/components/navbar.njk:32]: What does it mean by safe?

File: src/controllers/ProfileController.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+export const putProfilePicture = async (req: express.Request, res: express.Response) => {
+    
+    try {
+        await updateProfilePicture(req.body as ProfileRequest, req.session.token ?? '');
+        const profileUpdate: ProfileResponse = await getProfilePicture(req.session.token ?? '');
+    
+        if (profileUpdate.profilePicture) {
+            const scriptRegex = /<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>/gi;
```
Comment (Leebee1392) [inline src/controllers/ProfileController.ts:24]: Quick question, what is this regex actually doing?

File: src/controllers/AuthController.ts [hunk 6]
```diff
@@ -9,6 +11,15 @@ export const getLoginForm = (req: express.Request, res: express.Response) => {
 export const postLoginForm = async (req: express.Request, res: express.Response): Promise<void> => {
     try {
         req.session.token = await getToken(req.body as LoginRequest);
+        const profile: ProfileResponse = await getProfilePicture(req.session.token);
+        if(profile.profilePicture) {
+            const scriptRegex = /<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>/gi;
```
Comment (Leebee1392) [inline src/controllers/AuthController.ts:16]: Quick question, what is this regex actually doing?

File: views/profilePicture.njk [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+   })
+   .catch(err => {
+       console.log(err);
+   });
+
+document.getElementById('takePicture').addEventListener('click', takePicture);
+
+async function takePicture() {
+    var image = webcam.snap();
+    console.log(image);
```
Comment (Leebee1392) [inline views/profilePicture.njk:64]: Don't log the image

**Review by Leebee1392 — CHANGES_REQUESTED** [review]

File: src/scripts/imageScaling.js (hunk 1 above)
Comment (sdagn1) [reply src/scripts/imageScaling.js:21]: Nope, this has been corrected and committed.

File: views/layouts/main.njk (hunk 3 above)
Comment (sdagn1) [reply views/layouts/main.njk:19]: Yes, this essentially is chunk of nunjucks in the header that can be overwritten by other files. Without it there wouldn't be a way to put scripts in the header for a specific file.

File: views/components/navbar.njk (hunk 4 above)
Comment (sdagn1) [reply views/components/navbar.njk:32]: Discussed with Tom.

File: src/controllers/ProfileController.ts (hunk 5 above)
Comment (sdagn1) [reply src/controllers/ProfileController.ts:24]: Explained in person that it avoids script injection.

File: views/profilePicture.njk (hunk 7 above)
Comment (sdagn1) [reply views/profilePicture.njk]: This has been corrected and committed.

File: views/profilePicture.njk (hunk 2 above)
Comment (sdagn1) [reply views/profilePicture.njk]: This has been corrected and committed.

File: src/scripts/takePicture.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+       console.log(err);
+   });
+
+document.getElementById('takePicture').addEventListener('click', takePicture);
+
+async function takePicture() {
+    var image = webcam.snap();
+
+    var scaledImg = await resizedataURL(image, 32, 32);
+    console.log(scaledImg);
```
Comment (Leebee1392) [inline src/scripts/takePicture.js:24]: Remove console log

**Review by Leebee1392 — CHANGES_REQUESTED** [review]

File: src/scripts/takePicture.js (hunk 8 above)
Comment (sdagn1) [reply src/scripts/takePicture.js]: This has been corrected and committed.

**Review by Leebee1392 — APPROVED** [review]

**Review by Adnanhoq — APPROVED** [review]

**Review by Leebee1392 — APPROVED** [review]

**Review by AZZohaib — APPROVED** [review]

**Review by Leebee1392 — APPROVED** [review]

**Pull request merged**